thunderbirdparser/src/org/sleuthkit/autopsy/thunderbirdparser/EmailMessageThreader.java

```diff
@@ -0,0 +1,674 @@
+        // the real parent of container so throw away the old parent and set a
+        // new one.
+        if (container.hasParent()) {
+            container.getParent().removeChild(container);
+            container.setParent(null);
+        }
+
+        if (parent_ref != null) {
+            container.setParent(container);
+            parent_ref.addChild(container);
+        }
+    }
+
+    /**
+     * Creates a set of root container messages from the message-ID hashtable. A
+     * root Container is container that does not have a parent container.
+     *
+     * @param id_table Table of all known Containers
+     *
+     * @return A set of the root containers.
+     */
+    Set<Container> getRootSet(HashMap<?, Container> id_table) {
+        HashSet<Container> rootSet = new HashSet<>();
+
+        id_table.values().stream().filter((container)
+                -> (!container.hasParent())).forEachOrdered((container) -> {
+            rootSet.add(container);
+        });
+
+        return rootSet;
+    }
+
+    /**
+     * Remove Containers from containerSet if they do not have a message or
+     * children.
+     *
+     * @param containerSet A set of Container objects
+     */
+    void pruneEmptyContainers(Set<Container> containerSet) {
+        Set<Container> containersToRemove = new HashSet<>();
+        containerSet.forEach((container) -> {
+            if (!container.hasMessage() && !container.hasChildren()) {
+                containersToRemove.add(container);
+            } else {
+                pruneChildren(container);
+            }
+        });
+
+        containerSet.removeAll(containersToRemove);
+    }
+
+    /**
+     * Recursively work through the list of parent's children removing empy
+     * containers.
+     *
+     * @param parent
+     */
+    boolean pruneChildren(Container parent) {
+        if (parent == null) {
+            return false;
+        }
+
+        Set<Container> children = parent.getChildren();
+        Container grandParent = parent.getParent();
+
+        if (children == null) {
+            return false;
+        }
+
+        Set<Container> remove = new HashSet<>();
+        Set<Container> add = new HashSet<>();
+        for (Container child : parent.getChildren()) {
+            if (pruneChildren(child)) {
+                remove.add(child);
+                add.addAll(child.getChildren());
+                child.setParent(null);
+                child.clearChildren();
+
+            }
+        }
+
+        parent.addChildren(add);
+        parent.removeChildren(remove);
+
+        if (!parent.hasMessage() && grandParent != null) {
+            children.forEach((child) -> {
+                child.setParent(grandParent);
+            });
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Now that the emails are grouped together by references\message ID take
+     * another pass through and group together messages with the same simplified
+     * subject.
```

Comment:
For me, it would have been useful to have a line in here to the effect of "the same thread could currently be in spread out in different groups because some of the messages are missing.  Try to regroup them based on subject."